### PR TITLE
Implement Client Interfaces and Factory Methods

### PR DESCRIPTION
**Pull request by dvonthenen**
Addresses issue: https://github.com/deepgram/deepgram-dotnet-sdk/issues/273 

This implements client interfaces and factory methods to instantiate each of these clients. The point of this PR is encapsulate the clients without needing to expose the internal details.

Examples are provided for each client type.

### CODE REVIEW

**Review by SandraRodgers — DISMISSED** [review]

File: Deepgram/Clients/Live/v1/Client.cs [hunk 1]
```diff
@@ -2,17 +2,17 @@
...
 using Deepgram.Models.Live.v1;
+using Deepgram.Clients.Interfaces.v1;
 
 namespace Deepgram.Clients.Live.v1;
 
 /// <summary>
 /// Implements version 1 of the Live Client.
 /// </summary>
-public class Client : Attribute, IDisposable
+public class Client : Attribute, IDisposable, ILiveClient
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply Deepgram/Clients/Live/v1/Client.cs:15]: autocomplete! waves fist!

File: examples/analyze/factory_example/Program.cs [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+{
+    class Program
+    {
+        static async Task Main(string[] args)
+        {
+            // Initialize Library with default logging
+            // Normal logging is "Info" level
+            Library.Initialize();
+            // OR to set logging level
+            //Library.Initialize(LogLevel.Debug);
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply examples/analyze/factory_example/Program.cs:20]: this is intentional. ignoring suggestion.

File: examples/manage/factory_example/Factory.csproj [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+	<ItemGroup>	
+		<PackageReference Include="Microsoft.Extensions.Http" Version="8.0.0" />
+	</ItemGroup>
+	
+	<ItemGroup>
+	  <ProjectReference Include="..\..\..\Deepgram\Deepgram.csproj" />
+	</ItemGroup>
+
+    <ItemGroup>
+        <Using Include="Deepgram" />
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply examples/manage/factory_example/Factory.csproj:19]: this is intentional. ignoring suggestion.

File: examples/analyze/factory_example/Factory.csproj [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+	<ItemGroup>	
+		<PackageReference Include="Microsoft.Extensions.Http" Version="8.0.0" />
+	</ItemGroup>
+	
+	<ItemGroup>
+	  <ProjectReference Include="..\..\..\Deepgram\Deepgram.csproj" />
+	</ItemGroup>
+
+    <ItemGroup>
+	  <Using Include="Deepgram" />
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply examples/analyze/factory_example/Factory.csproj:19]: this is intentional. ignoring suggestion.

File: examples/analyze/factory_example/Factory.csproj [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+    </ItemGroup>
+
+	<ItemGroup>
+		<None Update="conversation.txt">
+		<CopyToOutputDirectory>Always</CopyToOutputDirectory>
+		</None>
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply examples/analyze/factory_example/Factory.csproj:25]: yes, it does! you need to recheck this.

File: examples/prerecorded/factory_example/Program.cs [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+{
+    class Program
+    {
+        static async Task Main(string[] args)
+        {
+            // Initialize Library with default logging
+            // Normal logging is "Info" level
+            //Library.Initialize();
+            // OR very chatty logging
+            Library.Initialize(LogLevel.Debug); // LogLevel.Default, LogLevel.Debug, LogLevel.Verbose
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply examples/prerecorded/factory_example/Program.cs:21]: this is intentional to easily switch to debug purposes

File: examples/prerecorded/factory_example/Factory.csproj [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+	</ItemGroup>
+
+    <ItemGroup>
+	  <Using Include="Deepgram" />
+    </ItemGroup>
+
+	<ItemGroup>
+		<None Update="Bueller-Life-moves-pretty-fast.wav">
+		<CopyToOutputDirectory>Always</CopyToOutputDirectory>
+		</None>
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply examples/prerecorded/factory_example/Factory.csproj:25]: the file does exist!

File: examples/manage/factory_example/Program.cs [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+{
+    class Program
+    {
+        static async Task Main(string[] args)
+        {
+            // Initialize Library with default logging
+            // Normal logging is "Info" level
+            Library.Initialize();
+            // OR very chatty logging
+            //Library.Initialize(LogLevel.Debug); // LogLevel.Default, LogLevel.Debug, LogLevel.Verbose
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply examples/manage/factory_example/Program.cs:20]: this is intentional to easily switch to debug purposes

File: examples/streaming/factory_example/Program.cs [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+                Model = "nova-2",
+                Encoding = "linear16",
+                SampleRate = 16000,
+                Punctuate = true,
+                SmartFormat = true,
+                InterimResults = true,
+                UtteranceEnd = "1000",
+                VadEvents = true,
+            };
+            await liveClient.Connect(liveSchema);
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply examples/streaming/factory_example/Program.cs:83]: this is intentional to easily switch to debug purposes

File: examples/streaming/factory_example/Program.cs [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+            microphone.Start();
+
+            // Wait for the user to press a key
+            Console.ReadKey();
+
+            // Stop the microphone
+            microphone.Stop();
+
+            // Stop the connection
+            await liveClient.Stop();
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply examples/streaming/factory_example/Program.cs:96]: this is just an example. this isnt needed.

File: Deepgram/Clients/Interfaces/v1/IAnalyzeClient.cs [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+    /// <summary>
+    /// Analyzes a file using the provided byte array and providing a CallBack
+    /// </summary>
+    /// <param name="source">file is the form of a byte[]</param>  
+    /// <param name="callBack">CallBack url</param>    
+    /// <param name="analyzeSchema">Options for the transcription<see cref="AnalyzeSchema"></param>
+    /// <returns><see cref="AsyncResponse"/></returns>
+    public Task<AsyncResponse> AnalyzeFileCallBack(byte[] source, string? callBack, AnalyzeSchema? analyzeSchema,
+        CancellationTokenSource? cancellationToken = default, Dictionary<string, string>? addons = null,
+        Dictionary<string, string>? headers = null);
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply Deepgram/Clients/Interfaces/v1/IAnalyzeClient.cs:50]: this is an invalid comment. not sure why this got generated.

File: Deepgram/Clients/Interfaces/v1/IPreRecordedClient.cs [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+    /// <summary>
+    /// Transcribe a file by providing a url and a CallBack
+    /// </summary>
+    /// <param name="source">Url to the file that is to be transcribed <see cref="UrlSource"/></param>
+    /// <param name="callBack">CallBack url</param>    
+    /// <param name="prerecordedSchema">Options for the transcription<see cref="PrerecordedSchema"></param>
+    /// <returns><see cref="AsyncResponse"/></returns>
+    public Task<AsyncResponse> TranscribeUrlCallBack(UrlSource source, string? callBack, PrerecordedSchema? prerecordedSchema,
+        CancellationTokenSource? cancellationToken = default, Dictionary<string, string>? addons = null,
+        Dictionary<string, string>? headers = null);
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply Deepgram/Clients/Interfaces/v1/IPreRecordedClient.cs:79]: the exception needs to be handled by the caller

File: Deepgram/Clients/Interfaces/v1/IAnalyzeClient.cs [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+    /// <summary>
+    /// Analyze a file by providing a url and a CallBack
+    /// </summary>
+    /// <param name="source">Url to the file that is to be analyzed <see cref="UrlSource"/></param>
+    /// <param name="callBack">CallBack url</param>    
+    /// <param name="analyzeSchema">Options for the transcription<see cref="AnalyzeSchema"></param>
+    /// <returns><see cref="AsyncResponse"/></returns>
+    public Task<AsyncResponse> AnalyzeUrlCallBack(UrlSource source, string? callBack, AnalyzeSchema? analyzeSchema,
+        CancellationTokenSource? cancellationToken = default, Dictionary<string, string>? addons = null,
+        Dictionary<string, string>? headers = null);
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply Deepgram/Clients/Interfaces/v1/IAnalyzeClient.cs:72]: the exception needs to be handled by the caller

File: examples/prerecorded/factory_example/Factory.csproj [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+	<ItemGroup>	
+		<PackageReference Include="Microsoft.Extensions.Http" Version="8.0.0" />
+	</ItemGroup>
+	
+	<ItemGroup>
+	  <ProjectReference Include="..\..\..\Deepgram\Deepgram.csproj" />
+	</ItemGroup>
+
+    <ItemGroup>
+	  <Using Include="Deepgram" />
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply examples/prerecorded/factory_example/Factory.csproj:19]: this is intentional and needed

File: examples/streaming/factory_example/Factory.csproj [hunk 15]
```diff
@@ -0,0 +1,19 @@
...
+    <Nullable>enable</Nullable>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <ProjectReference Include="..\..\..\Deepgram.Microphone\Deepgram.Microphone.csproj" />
+    <ProjectReference Include="..\..\..\Deepgram\Deepgram.csproj" />
+  </ItemGroup>
+  
+  <ItemGroup>
+    <Using Include="Deepgram" />
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply examples/streaming/factory_example/Factory.csproj:16]: this is intentional and needed

File: examples/analyze/factory_example/conversation.txt [hunk 16]
```diff
@@ -0,0 +1,71 @@
+Meet Deepgram Aura: real-time text-to-speech for real-time AI agents
+----------
+It’s been a year since large language models (LLMs) seemingly went mainstream overnight (Happy Birthday, ChatGPT!!!), and the world has witnessed both rapid development of these technologies and immense interest in their potential. We believe that we have reached an inflection point where voice-based interfaces will be the primary means to accessing LLMs and the experiences they unlock. Here are a few recent signals in support of our thesis:
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply examples/analyze/factory_example/conversation.txt:3]: this is just sample text

File: examples/analyze/factory_example/conversation.txt [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+
+- Good old fashioned voice notes are enjoying a healthy resurgence.
+
+- According to a recent survey, a majority of respondents stated phone calls are still their preferred communication channel for resolving customer service issues.
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply examples/analyze/factory_example/conversation.txt:7]: this is just sample text

File: examples/analyze/factory_example/conversation.txt [hunk 18]
```diff
@@ -0,0 +1,71 @@
...
+
+- An emerging boom in wearable devices equipped with continuous listening and speech AI technology is gaining steam.
+
+- OpenAI recently enabled voice interactions in ChatGPT.
+
+- A wave of interest in voice-first experiences and tools is sweeping across brands, investors, and tech companies.
+
+Thanks to ChatGPT and the advent of the LLM era, the conversational AI tech stack has advanced sufficiently to support productive (not frustrating) voice-powered AI assistants and agents that can interact with humans in a natural manner. We have already observed this from our most innovative customers who are actively turning to these technologies to build a diverse range of AI agents for voice ordering systems, interview bots, personal AI assistants, automated drive-thru tellers, and autonomous sales and customer service agents.
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply examples/analyze/factory_example/conversation.txt:15]: this is just sample text

File: examples/analyze/factory_example/conversation.txt [hunk 19]
```diff
@@ -0,0 +1,71 @@
...
+
+While these AI agents hold immense potential, many customers have expressed their dissatisfaction with the current crop of voice AI vendors, citing roadblocks related to speed, cost, reliability, and conversational quality. That’s why we’re excited to introduce our own text-to-speech (TTS) API, Deepgram Aura, built for real-time, conversational voice AI agents.
+
+Whether used on its own or in conjunction with our industry-leading Nova-2 speech-to-text API, we’ll soon provide developers with a complete speech AI platform, giving them the essential building blocks they need to build high throughput, real-time AI agents of the future.
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply examples/analyze/factory_example/conversation.txt:19]: this is just sample text

File: examples/analyze/factory_example/conversation.txt [hunk 20]
```diff
@@ -0,0 +1,71 @@
...
+
+On the flip side, High Throughput is about handling many quick, one-off interactions for real-time conversations at scale. Think fast food ordering, booking appointments, or inquiring about the latest deals at a car dealership. These tasks are relevant to just about everyone on the planet, and they require fast, efficient text-to-speech conversion for an AI agent to fulfill them. While voice quality is still important to keep users engaged, quality here is more about the naturalness of the flow of conversation and less about sounding like Morgan Freeman. But the primary focus for most customers in this category is on improving customer outcomes, meaning speed and efficiency are must-haves for ensuring these everyday exchanges are smooth and reliable at high volume.
+
+"Deepgram showed me less than 200ms latency today. That's the fastest text-to-speech I’ve ever seen. And our customers would be more than satisfied with the conversation quality."
+
+Jordan Dearsley, Co-founder at Vapi
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply examples/analyze/factory_example/conversation.txt:35]: this is just sample text

File: examples/analyze/factory_example/conversation.txt [hunk 21]
```diff
@@ -0,0 +1,71 @@
...
+
+
+More human than human
+----------
+With Aura, we’ll give realistic voices to AI agents. Our goal is to craft text-to-speech capabilities that mirror natural human conversations, including timely responses, the incorporation of natural speech fillers like 'um' and 'uh' during contemplation, and the modulation of tone and emotion according to the conversational context. We aim to incorporate laughter and other speech nuances as well. Furthermore, we are dedicated to tailoring these voices to their specific applications, ensuring they remain composed and articulate, particularly in enunciating account numbers and business names with precision.
+
+"I don’t really consider Azure and the other guys anymore because the voices sound so robotic."
+Jordan Dearsley, Co-founder at Vapi
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply examples/analyze/factory_example/conversation.txt:45]: this is just sample text

File: examples/analyze/factory_example/conversation.txt [hunk 22]
```diff
@@ -0,0 +1,71 @@
...
+
+And all the while, we’ve purposefully built our models under limited constraints to optimize their speed and efficiency. With support for dozens of languages and custom model training, our technical team has trained and deployed thousands of speech AI models (more than anybody else) which we operate and manage for our customers each day using our own computing infrastructure. 
+
+We also have our own in-house data labeling and data ops team with years of experience building bespoke workflows to record, store, and transfer vast amounts of audio in order to label it and continuously grow our bank of high-quality data (millions of hours and counting) used in our model training.
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply examples/analyze/factory_example/conversation.txt:58]: this is just sample text

File: examples/analyze/factory_example/conversation.txt [hunk 23]
```diff
@@ -0,0 +1,71 @@
...
+
+"Deepgram is a valued partner, providing our customers with high throughput speech-to-text that delivers unrivaled performance without tradeoffs between quality, speed, and cost. We're excited to see Deepgram extend their speech AI platform and bring this approach to the text-to-speech market." - Richard Dumas, VP AI Product Strategy at Five9
+
+
+What's Next
+----------
+As we’ve discussed, scaled voice agents are a high throughput use case, and we believe their success will ultimately depend on a unified approach to audio, one that strikes the right balance between natural voice quality, responsiveness, and cost-efficiency. And with Aura, we’re just getting started. We’re looking forward to continuing to work with customers like Asurion and partners like Five9 across speech-to-text AND text-to-speech as we help them define the future of AI agents, and we invite you to join us on this journey.
+
+We expect to release generally early next year, but if you’re working on any real-time AI agent use cases, join our waitlist today to jumpstart your development in production as we continue to refine our model and API features with your direct feedback.
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply examples/analyze/factory_example/conversation.txt:71]: this is just sample text

File: examples/analyze/factory_example/conversation.txt [hunk 24]
```diff
@@ -0,0 +1,71 @@
...
+
+- OpenAI recently enabled voice interactions in ChatGPT.
+
+- A wave of interest in voice-first experiences and tools is sweeping across brands, investors, and tech companies.
+
+Thanks to ChatGPT and the advent of the LLM era, the conversational AI tech stack has advanced sufficiently to support productive (not frustrating) voice-powered AI assistants and agents that can interact with humans in a natural manner. We have already observed this from our most innovative customers who are actively turning to these technologies to build a diverse range of AI agents for voice ordering systems, interview bots, personal AI assistants, automated drive-thru tellers, and autonomous sales and customer service agents.
+
+While these AI agents hold immense potential, many customers have expressed their dissatisfaction with the current crop of voice AI vendors, citing roadblocks related to speed, cost, reliability, and conversational quality. That’s why we’re excited to introduce our own text-to-speech (TTS) API, Deepgram Aura, built for real-time, conversational voice AI agents.
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply examples/analyze/factory_example/conversation.txt:17]: this is just sample text

File: examples/analyze/factory_example/conversation.txt [hunk 25]
```diff
@@ -0,0 +1,71 @@
+Meet Deepgram Aura: real-time text-to-speech for real-time AI agents
+----------
+It’s been a year since large language models (LLMs) seemingly went mainstream overnight (Happy Birthday, ChatGPT!!!), and the world has witnessed both rapid development of these technologies and immense interest in their potential. We believe that we have reached an inflection point where voice-based interfaces will be the primary means to accessing LLMs and the experiences they unlock. Here are a few recent signals in support of our thesis:
+
+- Good old fashioned voice notes are enjoying a healthy resurgence.
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply examples/analyze/factory_example/conversation.txt:5]: this is just sample text

File: examples/analyze/factory_example/conversation.txt [hunk 26]
```diff
@@ -0,0 +1,71 @@
+Meet Deepgram Aura: real-time text-to-speech for real-time AI agents
```
[the remark this answers is not in the document]
Comment (dvonthenen) [reply examples/analyze/factory_example/conversation.txt:1]: this is just sample text

**Review by SandraRodgers — APPROVED** [review]

**Pull request merged**